src/surge-xt/gui/widgets/OscillatorWaveformDisplay.cpp

```diff
@@ -1267,260 +1268,276 @@ struct WaveTable3DEditor : public juce::Component,
+        }
+        return false;
+    }
 
-        // Now we have a sort of skew back and offset as we go. The skew is sort of a rotation
-        // and the depth is sort of how flattened it is. Finally the hCompress augments height.
-        auto skewPct = 0.4;
-        auto depthPct = 0.6;
-        auto hCompress = 0.55;
+    void resized() override { backingImage = nullptr; }
 
-        // calculate thinning factor for frame drawing
-        int thintbl = 1;
-        int nt = wt.n_tables;
+    /*
+        create a interpolated frame based on a float frame value
+    */
 
-        while (nt > 16)
+    void getInterpFrame(float *samples, float frame, float frames, bool processForReal,
+                        bool useCache)
+    {
+        auto &wt = oscdata->wt;
+        float actualFrame = (frame / (frames - 1.f)) * (float)(wt_nframes - 1);
+        int frameFrom = (int)(actualFrame);
+        float proc = actualFrame - (float)frameFrom;
+
+        if (processForReal == false)
         {
-            thintbl <<= 1;
-            nt >>= 1;
-        }
+            if (samplesCached[(int)frame][0] == 0.f || useCache == false)
+            {
+                int frameTo = frameFrom + 1 > wt_nframes - 1 ? frameFrom : frameFrom + 1;
 
-        // calculate thinning factor for sample drawing
-        int thinsmp = 1;
-        int s = smp;
+                for (int i = 0; i < rendered_samples; i++)
+                {
 
-        while (s > 128)
-        {
-            thinsmp <<= 1;
-            s >>= 1;
+                    int pos = floor((((float)i) / (float)std::max(rendered_samples - 1, 1)) *
+                                    ((float)wt_size - 1.f));
+
+                    samples[i] = wt.TableF32WeakPointers[0][frameFrom][pos] * (1.f - proc) +
+                                 wt.TableF32WeakPointers[0][frameTo][pos] * proc;
+                    if (useCache)
+                        samplesCached[(int)frame][i + 1] = samples[i];
+                }
+
+                samplesCached[(int)frame][0] = 1.f;
+            }
+            else
+            {
+                for (int i = 0; i < rendered_samples; i++)
+                {
+                    samples[i] = samplesCached[(int)frame][i + 1];
+                }
+            }
         }
+    }
+
+    void drawWavetable(juce::Graphics &g)
+    {
 
-        static constexpr float backingScale = 2.f;
+        auto osc = parent->setupOscillator();
+        osc->init(0, true, false);
 
-        auto wxf = w;
-        auto hxf = h;
+        float opacity = 1;
 
-        if (!backingImage)
+        float opacityMultiplier = parent->isMuted ? 0.5f : 1.f;
+
+        /*
+            Draw the base layer
+        */
+
+        if (paramsHaveChanged() || !backingImage)
         {
+
             backingImage =
-                std::make_unique<juce::Image>(juce::Image::PixelFormat::ARGB, wxf, hxf, true);
+                std::make_unique<juce::Image>(juce::Image::PixelFormat::ARGB, w * 2, h * 2, true);
```

Comment:
This is more the line that gives me concern. Why do we have to recreate the backing image on every param change? Can't we simply clear it if it exists? This is quite a bit of allocation.

Also w and h are virtual pixel sizes. Is the *2 for retina? The prior code was path based so did fine at 400% on a retina display. Will this now be pixelated in that case?

Reply:
ah you are absolutely right. That was just how the code was before. It got nulled and recreated on resize, so I just kept it that way as I thought there was a good reason for that. 
I increased the resolution because it was pixelated on my computer atleast, but I can try and reduce it back to see if that is enough. I doubt it though.

Reply:
![image](https://github.com/user-attachments/assets/c72cee42-2e21-4d5e-9ee3-e99f1410c58a)
difference between 1x vs 2x

Reply:
is that zoom factor accessible? I could just read that in the resize function och recreate the image only when its changed

Reply:
Yeah my concern wasn't that 2x was too high. It was that 2x was to low! The w and h in this case are in unzoomed pixels 

and yes getZoomFactor * 2 will be a good answer

Reply:
oh the units on getzoomfactor may be 100, 125, 150 not 1.0, 1.25, 1.50 just so you know. I never remember until I either get a tiny UI or an out of memory. But just a note.